Fix GCC 11 Compiler Warnings

franzpoeschel opened the pull request:
1. Compile warnings in Attribute.hpp:
    g++ 11 does not understand that these variables are not primitive types
    and complains about possibly uninitialized values. Use the {}
    constructor to fix this.
2. We still had two instances of AccessType in the tests, replace this with Access

franzpoeschel committed:
Compile warnings in Attribute.hpp

G++ 11 does not understand that these variables are not primitive types
and complains about possibly uninitialized values. Use the {}
constructor to fix this.

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -421,8 +421,7 @@ TEST_CASE("multiple_series_handles_test", "[serial]")
         std::unique_ptr<openPMD::Series> series_ptr;
         {
-            openPMD::Series series(
-                "sample%T.json", openPMD::AccessType::CREATE);
```
We left those intentionally to test that the old type still works.

We could silence this at one last location with a local suppressor?

franzpoeschel replied:
I was under the impression that they sneaked in because this test comes from a long-standing pull request (the bug tested by this took a long time to fix).
But we can definitely use the chance to test for existence of AccessType with this

ax3l commented:
PR rebased.

franzpoeschel committed:
Ignore deprecated AccessType in SerialIOTest

franzpoeschel committed:
ICC issue: statement unreachable

franzpoeschel commented on include/openPMD/backend/Attribute.hpp:
```diff
@@ -199,7 +200,6 @@ auto doConvert(T *pv) -> U
     }
 
-    (void)pv;
     throw std::runtime_error("getCast: no cast possible.");
```
Need to put these into an else branch, ICC seems to determine unreachable statements after resolving if constexpr

ax3l replied:
ah classic. a range of NVCC 11 minor releases does that as well

franzpoeschel committed:
Put throw statement into an else branch

ICC seems to resolve reachability of statements after resolving
constexpr if

franzpoeschel committed:
Add pragma for ICPC

franzpoeschel committed:
Push pragma only in G++ and Clang

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -414,6 +414,16 @@ TEST_CASE("available_chunks_test_json", "[serial][json]")
+    /*
+     * clang also understands these pragmas.
+     */
+#if defined(__GNUC__) || defined(__clang__)
```
`__GNUC__` is quite often set (GCC, Clang, ICC, etc.)
but the version is specific to GCC:
https://gcc.gnu.org/onlinedocs/cpp/Common-Predefined-Macros.html

ax3l commented on include/openPMD/backend/Attribute.hpp:
```diff
@@ -180,27 +185,42 @@ auto doConvert(T *pv) -> U
+ * then complain about unreachable code.
+ * https://community.intel.com/t5/Intel-C-Compiler/quot-if-constexpr-quot-and-quot-missing-return-statement-quot-in/td-p/1154551
+ */
+#pragma warning(disable : 1011)
```
Since this is a `.hpp` file, do you need to `pop` this again?

franzpoeschel replied:
Yes, I need to reset it to default again after this

ax3l replied:
done in latest commit :)

franzpoeschel committed:
Two fixes

1. Reset to default warning behavior after disabling
2. Use __GNUC_MINOR__ instead of __GNUC__

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -414,6 +414,13 @@ TEST_CASE("available_chunks_test_json", "[serial][json]")
+     */
+#if defined(__GNUC__MINOR__) || defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
```
With this one set, you can remove `-Wno-deprecated-declarations` from quite a few CI scripts in `.github/workflows/`

ax3l replied:
-> #1246

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -414,6 +414,13 @@ TEST_CASE("available_chunks_test_json", "[serial][json]")
+    /*
+     * clang also understands these pragmas.
+     */
+#if defined(__GNUC__MINOR__) || defined(__clang__)
```
Typo fixed in #1255: `__GNUC_MINOR__`

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -456,6 +463,12 @@ TEST_CASE("multiple_series_handles_test", "[serial]")
+#if defined(__INTEL_COMPILER)
+#pragma warning(disable : 2282)
+#endif
+#if defined(__GNUC__MINOR__) || defined(__clang__)
```
same here

ax3l commented:
I have problems applying this and the following PRs to the `release-0.14.5` branch

ax3l commented:
Applied to `0.14.5` via https://github.com/ax3l/openPMD-api/pull/6 :tada:
